lib/core/buffer.c

```diff
@@ -842,6 +840,41 @@ void swap_frames(struct Buffer* from_buf, int from_frame_id, struct Buffer* to_b
     to_buf->frames[to_frame_id] = temp_frame;
 }
 
+void safe_swap_frame(struct Buffer* src_buf, int src_frame_id, struct Buffer* dest_buf,
+                     int dest_frame_id) {
+    assert(src_buf != dest_buf);
+    assert(src_buf != NULL);
+    assert(dest_buf != NULL);
+    assert(src_frame_id >= 0);
+    assert(src_frame_id < src_buf->num_frames);
+    assert(dest_frame_id >= 0);
+    assert(dest_frame_id < dest_buf->num_frames);
+
+    // Buffer sizes must match exactly
+    if (src_buf->frame_size != dest_buf->frame_size) {
+        FATAL_ERROR_F("Buffer sizes must match for direct copy (%s.frame_size != %s.frame_size)",
+                      src_buf->buffer_name, dest_buf->buffer_name);
+    }
+
+    if (get_num_producers(dest_buf) > 1) {
+        FATAL_ERROR_F("Cannot swap/copy frames into dest buffer %s with more than one producer",
+                      dest_buf->buffer_name);
+    }
+
+    int num_consumers = get_num_consumers(src_buf);
+
+    // Copy or transfer the data part.
+    if (num_consumers == 1) {
+        // Swap the frames
+        uint8_t* temp_frame = src_buf->frames[src_frame_id];
+        src_buf->frames[src_frame_id] = dest_buf->frames[dest_frame_id];
```

Comment:
Because if there is more than one consumer of this buffer, then this frame might not yet have been read by one of them.  So the swap would result it the other consumers possibly getting the wrong data.   With the deep copy below however this isn't a problem, but it isn't as fast.